File: medium/3020_find_the_maximum_number_of_elements_in_subset/solution.cpp
#include <unordered_map>
#include <vector>

class Solution {
public:
  int maximumLength(std::vector<int> &nums) {
    std::unordered_map<int, int> numsCounter;
    for (const int &num : nums)
      ++numsCounter[num];

    int maxLen = 0;
    if (numsCounter.find(1) != numsCounter.end()) {
      maxLen = numsCounter[1];
      if (maxLen % 2 == 0)
        --maxLen;

      numsCounter.erase(1);
    }

    for (const auto &pair : numsCounter) {
      int num = pair.first;
      int currLen = 0;
      auto it = numsCounter.find(num);

      while (it != numsCounter.end() && it->second >= 2) {
        currLen += 2;

        if (num > INT_MAX / num) {
          it = numsCounter.end();
          break;
        }

        num *= num;
        it = numsCounter.find(num);
      }

      it = numsCounter.find(num);
      if (it != numsCounter.end() && it->second == 1)
        maxLen = std::max(maxLen, currLen + 1);
      else
        maxLen = std::max(maxLen, currLen - 1);
    }

    return maxLen;
  }
};

File: medium/3020_find_the_maximum_number_of_elements_in_subset/solution.py
from collections import Counter


class Solution:
    def maximumLength(self, nums: list[int]) -> int:
        numsCounter = Counter(nums)

        if 1 in numsCounter:
            maxLen = numsCounter.pop(1)
            if maxLen % 2 == 0:
                maxLen -= 1
        else:
            maxLen = 0

        for num in numsCounter.keys():
            currLen = 0

            while numsCounter.get(num, 0) >= 2:
                num *= num
                currLen += 2

            if numsCounter.get(num, 0) == 1:
                maxLen = max(maxLen, currLen + 1)
            else:
                maxLen = max(maxLen, currLen - 1)

        return maxLen
